src/Utilities/MoistThermodynamics/src/MoistThermodynamics.jl

```diff
@@ -462,49 +784,212 @@ function saturation_adjustment(e_int, ρ, q_tot, T_init = T_triple)
     return air_temperature(e_int, q_tot, q_liq, q_ice)
   end
 end
+function saturation_adjustment_q_t_θ_l(θ_liq::DT, q_tot::DT, ρ::DT, p::DT) where DT
+  T_1 = θ_liq*(p/DT(MSLP))^(DT(R_d)/DT(cp_d))
+  qv_star = saturation_shum(T_1, ρ)
+  if (q_tot <= qv_star) # If not saturated
+    return T_1
+  else  # If not saturated, iterate
+    T_2 = T_1 + (q_tot - qv_star) * latent_heat_vapor(T_1) /((1-q_tot)*DT(cp_d) + qv_star * DT(cp_v))
+    function eos_root(T, args)
+      qv_star = saturation_shum(T, ρ)
+      θ = T*(DT(MSLP) / p)^(DT(R_d)/DT(cp_d))
+      temp = exp(-latent_heat_vapor(T)/(T*DT(cp_d))*(q_tot - qv_star)/(1-q_tot))
+      return θ_liq - θ*temp
+    end
+    T, converged = find_zero(eos_root, T_1, T_2, Tuple(1),
+                             IterParams(DT(1e-3), 10),
+                             SecantMethod())
+    return T
+  end
+end
 
 """
+    liquid_ice_pottemp(ts::ThermodynamicState)
     liquid_ice_pottemp(T, p[, q_tot=0, q_liq=0, q_ice=0])
 
-Return the liquid-ice potential temperature, given the temperature `T`,
-pressure `p`, total specific humidity `q_tot`, liquid specific humidity `q_liq`,
-and ice specific humidity `q_ice`.
+Return the liquid-ice potential temperature,
+given a thermodynamic state `ts` or
+
+ - `T` temperature
+ - `p` pressure
+and, optionally,
+ - `q_tot` total specific humidity
+ - `q_liq` liquid specific humidity
+ - `q_ice` ice specific humidity
 """
-function liquid_ice_pottemp(T, p, q_tot=0, q_liq=0, q_ice=0)
+function liquid_ice_pottemp(T::DT, p::DT, q_tot::DT=DT(0), q_liq::DT=DT(0), q_ice::DT=DT(0)) where DT
 
     # isobaric specific heat of moist air
     _cp_m   = cp_m(q_tot, q_liq, q_ice)
 
     # liquid-ice potential temperature, approximating latent heats
     # of phase transitions as constants
-    return dry_pottemp(T, p, q_tot, q_liq, q_ice) * exp(-(LH_v0*q_liq + LH_s0*q_ice)/(_cp_m*T))
+    return dry_pottemp(T, p, q_tot, q_liq, q_ice) * exp(-(DT(LH_v0)*q_liq + DT(LH_s0)*q_ice)/(_cp_m*T))
+
+end
+function liquid_ice_pottemp(ts::ThermodynamicState)
+
+    return liquid_ice_pottemp(air_temperature(ts),
+                              air_pressure(ts),
+                              ts.q_tot,
+                              phase_partitioning_eq(ts)...)
+
+end
+
+"""
```

Comment:
Please remove the mixing ratios, unless there is a strong reason why we need them. We should stay within one set of moisture variables to avoid unnecessary confusion, and I right now I see no reason why we would need the mixing ratios.

Reply:
They'd be convenient. I can work around omitting them out for now, but I'll probably write the `tke` buoyancy source in terms of mixing ratios in the short-term EDMF implementation.

Reply:
Let's talk about the buoyancy source terms. They should be written in terms of the specific humidities we have, not mixing ratios.